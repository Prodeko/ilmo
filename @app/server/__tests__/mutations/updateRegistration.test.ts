import {
  asRoot,
  createEventDataAndLogin,
  deleteTestData,
  runGraphQLQuery,
  sanitize,
  setup,
  teardown,
} from "../helpers";

beforeEach(deleteTestData);
beforeAll(setup);
afterAll(teardown);

const updateRegistrationMutation = `
mutation UpdateEventRegistration(
  $updateToken: String!
  $firstName: String!
  $lastName: String!
) {
  updateRegistration(
    input: {
      updateToken: $updateToken
      firstName: $firstName
      lastName: $lastName
    }
  ) {
    registration {
      id
      fullName
    }
  }
}`;

describe("UpdateRegistration", () => {
  it("can update a registration with a valid updateToken", async () => {
    const { registrationSecrets } = await createEventDataAndLogin();
    const { update_token: updateToken } = registrationSecrets[0];

    await runGraphQLQuery(
      updateRegistrationMutation,

      // GraphQL variables:
      {
        updateToken,
        firstName: "P채ivi",
        lastName: "Tetty",
      },

      // Additional props to add to `req` (e.g. `user: {session_id: '...'}`)
      {
        user: {},
      },

      // This function runs all your test assertions:
      async (json, { pgClient }) => {
        expect(json.errors).toBeFalsy();
        expect(json.data).toBeTruthy();

        const updatedRegistration = json.data!.updateRegistration.registration;
        expect(updatedRegistration).toBeTruthy();

        expect(sanitize(updatedRegistration)).toMatchInlineSnapshot(`
          Object {
            "fullName": "P채ivi Tetty",
            "id": "[id-1]",
          }
        `);

        const { rows } = await asRoot(pgClient, () =>
          pgClient.query(
            `SELECT * FROM app_public.registrations WHERE id = $1`,
            [updatedRegistration.id]
          )
        );

        if (rows.length !== 1) {
          throw new Error("Registration not found!");
        }
        expect(rows[0].id).toEqual(updatedRegistration.id);
        expect(rows[0].first_name).toEqual("P채ivi");
        expect(rows[0].last_name).toEqual("Tetty");
      }
    );
  });

  it("can't update registration if registration token is not valid", async () => {
    await runGraphQLQuery(
      updateRegistrationMutation,

      // GraphQL variables:
      {
        // Invalid updateToken
        updateToken: "a7def7b2-1687-48d8-839e-55e57f6ade85",
        firstName: "P채ivi",
        lastName: "Tetty",
      },

      // Additional props to add to `req` (e.g. `user: {session_id: '...'}`)
      {},

      // This function runs all your test assertions:
      async (json) => {
        expect(json.errors).toBeTruthy();

        const message = json.errors![0].message;
        const code = json.errors![0].extensions.exception.code;
        expect(message).toEqual("Registration matching token was not found.");
        expect(code).toEqual("NTFND");
      }
    );
  });
});
