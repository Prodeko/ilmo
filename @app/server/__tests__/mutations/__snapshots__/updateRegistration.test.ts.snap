// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateRegistration can update a registration with a valid updateToken 2`] = `
Object {
  "data": Object {
    "updateRegistration": Object {
      "registration": Object {
        "fullName": "PÃ¤ivi Tetty",
        "id": "[id-1]",
      },
    },
  },
}
`;

exports[`UpdateRegistration can't update registration if registration token is not valid 1`] = `
Object {
  "data": Object {
    "updateRegistration": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "exception": Object {
          "code": "NTFND",
          "severity": "ERROR",
          "where": "PL/pgSQL function app_public.update_registration(text,text,text) line 11 at RAISE",
        },
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 7,
        },
      ],
      "message": "Registration matching token was not found.",
      "path": Array [
        "updateRegistration",
      ],
    },
  ],
}
`;
